poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+/* Unmarshal a PromiseSignature struct using an io.Reader
+ *
+ * Arguments
+ *    r = the reader to use for unmarshalling
+ *
+ * Returns
+ *   The number of bytes read
+ *   The error status of the read (nil if no errors)
+ */
+func (p *PromiseSignature) UnmarshalFrom(r io.Reader) (int, error) {
+	// Retrieve the signature length from the reader
+	buf := make([]byte, uint32Size)
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+
+	sigLen := int(binary.LittleEndian.Uint32(buf))
+
+	// Calculate the length of the entire message and create the new buffer.
+	finalBuf := make([]byte, uint32Size+sigLen)
+
+	// Copy the old buffer into the new
+	copy(finalBuf, buf)
+
+	// Read the rest and unmarshal.
+	m, err := io.ReadFull(r, finalBuf[n:])
+	if err != nil {
+		return n + m, err
+	}
+	return n + m, p.UnmarshalBinary(finalBuf)
+}
+
+/* Returns a string representation of the PromiseSignature for easy debugging
+ *
+ * Returns
+ *   The PromiseSignature's string representation
+ */
+func (p *PromiseSignature) String() string {
+	s := "{PromiseSignature:\n"
+	s += "Suite => " + p.suite.String() + ",\n"
+	s += "Signature => " + hex.EncodeToString(p.signature) + "\n"
+	s += "}\n"
+	return s
+}
+
+/* The BlameProof struct provides an accountability measure. If a promiser
+ * decides to construct a faulty share, insurers can construct a BlameProof
+ * to show that the promiser is malicious.
+ *
+ * The insurer provides the Diffie-Hellman shared secret with the promiser so
+ * that others can decode the share in question. A zero knowledge proof is
+ * provided to prove that the shared secret was constructed properly. Lastly, a
+ * PromiseSignature is attached to prove that the insurer endorses the blame.
+ * When other servers receive the BlameProof, they can then verify whether the
+ * promiser is malicious or the insurer is falsely accusing the promiser.
+ *
+ * To quickly summarize the blame procedure, the following must hold for the
+ * blame to succeed:
+ *
+ *   1. The PromiseSignature must be valid
+ *
+ *   2. The Diffie-Hellman key must be verified to be correct
+ *
+ *   3. The insurer's share when decrypted must fail the PubPoly.Check of
+ *   the Promise struct
+ *
+ * If all hold, the promiser is proven malicious. Otherwise, the insurer is
+ * slanderous.
+ *
+ * Beyond unmarshalling, users of this code need not worry about constructing a
+ * BlameProof struct themselves. The Promise struct knows how to create a
+ * BlameProof via the Promise.Blame method.
+ */
+
+type BlameProof struct {
+
+	// The suite used throughout the BlameProof
+	suite abstract.Suite
+
+	// The Diffie-Hellman shared secret between the insurer and promiser
+	diffieKey abstract.Point
+
+	// A HashProve proof that the insurer properly constructed the Diffie-
+	// Hellman shared secret
+	diffieKeyProof []byte
```

Comment:
In my revisions of creating a PromiseResponse, I no longer need to make BlameProof public. The name "proof" would conflict with crypto/proof, so I named the struct blameProof.

However, I changed diffieKeyProof  to proof.